aten/src/ATen/native/cuda/Blas.cpp

```diff
@@ -168,25 +169,50 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
             c10::nullopt /* pin_memory */));
   }
 
-  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!result_->is_conj());
-
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, scalar_type, "addmm_cuda", [&] {
-    scalar_t alpha_val = alpha.to<scalar_t>();
-    scalar_t beta_val = beta.to<scalar_t>();
-    scalar_t* mat1_ptr = mat1_->data_ptr<scalar_t>();
-    scalar_t* mat2_ptr = mat2_->data_ptr<scalar_t>();
-    scalar_t* result_ptr = result_->data_ptr<scalar_t>();
-    at::cuda::blas::gemm<scalar_t>(
-      transpose_mat1 ? mat1_->is_conj() ? 'c' : 't' : 'n',
-      transpose_mat2 ? mat2_->is_conj() ? 'c' : 't' : 'n',
-      m, n, k,
-      alpha_val,
-      mat1_ptr, mat1_ld,
-      mat2_ptr, mat2_ld,
-      beta_val,
-      result_ptr, result_ld
+  // TODO: is it possible to combine the dispatch code?
```

Comment:
I think so, as long as you call `integer_gemm` simply `gemm`. Since integer tensors do not have a float bit, the handling of `'c'` should be fine.

My only complaint would be that the function would live in the `blas::` namespace, even though it's not a `cuBLAS` function. This is a bit annoying, but it's not terribly bad.

Reply:
It does work:
```c++
#include <iostream>

template<typename scalar_t>
scalar_t f1(scalar_t x) {
  std::cout << "calling f1" << std::endl;
  return x;
}

template<typename scalar_t>
scalar_t f2(scalar_t x) {
  std::cout << "calling f2" << std::endl;
  return x;
}

template<typename scalar_t>
struct F {
  static decltype(&f1<scalar_t>) f(scalar_t x) {
    if (x < 3) return f1<scalar_t>;
    return f2<scalar_t>;
  }
};

int main(int argn, char* args[]) {
  F<int>::f(2)(3);
  F<int>::f(5)(4);
}

```
Outputs 
```
calling f1
calling f2

```

Reply:
Oh, I see what you mean now. Yes, you are right, this works. You could do something like:
```c++
const bool is_int = scalar_type == ScalarType::Long || scalar_type == ScalarType::Int;

auto gemm = is_int ? &at::cuda::blas::integer_gemm<scalar_t> 
                                : &at::cuda::blas::gemm<scalar_t>;
```

Reply:
How do you change the `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2` macro and other macros like it so that they can dispatch to integer types too? From what I can see you still need to have separate FP/complex+FP16 macros along with an extra call to `AT_DISPATCH_INDEX_TYPES` if you want to dispatch to `int32` and `int64`.

The above solution would reduce duplication in terms of code within the macros but I'll still need to call two separate macros.

Reply:
I bet there must dispatch keys for almost all types, I guess. You can check indexing operators as they work for all dtypes.

Reply:
I believe that dispatch at floating/complex and two more that you specify should do the job.

Reply:
Its already using B16 as the two more data types.

Reply:
there must AND3 macro, right?
```
pytorch-cuda-dev) nik@qgpu3:~/git/Quansight/pytorch$ git grep AND3                                                                                                                                               
/usr/bin/pager: /home/nik/.conda/envs/pytorch-cuda-dev/lib/libtinfo.so.6: no version information available (required by /usr/bin/pager)                                                                           
aten/src/ATen/Dispatch.h:#define AT_DISPATCH_ALL_TYPES_AND3(                                         \                                                                                                            
aten/src/ATen/Dispatch.h:#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                             \                                                                                                            
aten/src/ATen/ScalarOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                             
aten/src/ATen/Utils.cpp:AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)                                                                                                                                 
aten/src/ATen/native/BinaryOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                      
aten/src/ATen/native/Copy.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kHalf, kBool, kBFloat16, self.scalar_type(), "copy_", [&] {                                                                                
aten/src/ATen/native/ReduceOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, iter.input_dtype(), "equal_cpu", [&] {
aten/src/ATen/native/Scalar.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16,                                                              
aten/src/ATen/native/TensorAdvancedIndexing.cpp:      AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16,                                                            
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorFactories.cpp:AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, DEFINE_CAST_OP)      
...
```

Reply:
Hmm, indeed, there is no such macro. What about writing a new version with AND3?

Reply:
We'll need AND4 (int32 + int64 + Bfloat16 + FP16). Is that justified for not having a single if-else in only one part of the code?

Reply:
Well, we can indeed keep it in an if-else part but it is still possible to compress the logic in a single function, right?
Something like
```
template <typename scalar_t>
void logic(...)

if (...) {
  AT_DISPATCH_FLOATING_AND_COMPLEX(..., logic<scalar_t>());
}
else {
 AT_DISPATCH_INDEX_TYPE(..., logic<scalar_t>());
}
```

Reply:
I actually wonder why AT_DISPATCH is not implemented with variadics...